bokeh/util/api_crawler.py

```diff
@@ -0,0 +1,232 @@
+import ast, os, copy
+
+__all__ = ["api_crawler", "differ"]
+
+
+class APICrawler(object):
+    """
+    Api Crawler: Given a directory, it will use AST to crawl through each of the
+    File and produce AST nodes, which can then be parsed into a representation
+    of the structure of that file. The execution...
+    """
+
+    # Exclude these folders from search.
+    exclude = ("tests", "static", "sampledata", "mplexplorer")
+
+
+    # Point the crawler at this directory.
+    def __init__(self, directory):
+        self.directory = directory
+
+    def is_public(self, name):
+        # Determines whether a given function or class is public or private. Names
+        # with underscores are not allowed unless the name is "__init__"
+        return not name.startswith("_", 0, 1) or name == "__init__"
+
+    def is_function(self, ast_node):
+        # Checks whether a node is a top level function. Checks whether it is an AST
+        # function instance. Also uses column offset, i.e. whether the function has
+        # any tabs or spaces before it.
+        return isinstance(ast_node, ast.FunctionDef) and ast_node.col_offset == 0
```

Comment:
I presume there is no better way of doing this, without actually interpreting the code? This can be easily fooled by top-level `try-catch` or `if-else`, though this doesn't happen at all (I presume).